modules/objdetect/test/test_qrcode_encode.cpp

```diff
@@ -0,0 +1,257 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "test_precomp.hpp"
+namespace opencv_test { namespace {
+
+std::string encode_qrcode_images_name[] = {
+        "version1_mode1.png", "version1_mode2.png", "version1_mode4.png",
+        "version2_mode1.png", "version2_mode2.png", "version2_mode4.png",
+        "version3_mode2.png", "version3_mode4.png",
+        "version4_mode4.png"
+};
+
+const Size fixed_size = Size(600, 600);
+const float border_width = 2.0;
+int countDiffPixels(cv::Mat in1, cv::Mat in2) {
+    cv::Mat diff;
+    cv::compare(in1, in2, diff, cv::CMP_NE);
+    return cv::countNonZero(diff);
+}
+int establishCapacity(int mode, int version, int capacity)
+{
+    int result = 0;
+    capacity *= 8;
+    capacity -= 4;
+    switch (mode)
+    {
+        case QR_MODE_NUM:
+        {
+            if (version >= 10)
+                capacity -= 12;
+            else
+                capacity -= 10;
+            int tmp = capacity / 10;
+            result = tmp * 3;
+            if (tmp * 10 + 7 <= capacity)
+                result += 2;
+            else if (tmp * 10 + 4 <= capacity)
+                result += 1;
+            break;
+        }
+        case QR_MODE_ALPHA:
+        {
+            if (version < 10)
+                capacity -= 9;
+            else
+                capacity -= 13;
+            int tmp = capacity / 11;
+            result = tmp * 2;
+            if (tmp * 11 + 6 <= capacity)
+                result++;
+            break;
+        }
+        case QR_MODE_BYTE:
+        {
+            if (version > 9)
+                capacity -= 16;
+            else
+                capacity -= 8;
+            result = capacity / 8;
+            break;
+        }
+    }
+    return result;
+}
+
+// #define UPDATE_TEST_DATA
+#ifdef UPDATE_TEST_DATA
+
+TEST(Objdetect_QRCode_Encode, generate_test_data)
+{
+    const std::string root = "qrcode/encode";
+    const std::string dataset_config = findDataFile(root +"/"+ "dataset_config.json");
+    FileStorage file_config(dataset_config, FileStorage::WRITE);
+
+    file_config << "test_images" << "[";
+    size_t images_count = sizeof(encode_qrcode_images_name) / sizeof(encode_qrcode_images_name[0]);
+    for (size_t i = 0; i < images_count; i++)
+    {
+        file_config << "{:" << "image_name" << encode_qrcode_images_name[i];
+        std::string image_path = findDataFile(root +"/"+ encode_qrcode_images_name[i]);
+
+        /**read from test set*/
+        Mat src = imread(image_path, IMREAD_GRAYSCALE), straight_barcode;
```

Comment:
> straight_barcode

Avoid multiple declarations on the single line (allowed for simple cases without initialization only).
Must be defined near its usage.